src/Kestrel.Transport.Sockets/Internal/CoalescingScheduler.cs

```diff
@@ -0,0 +1,84 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.IO.Pipelines;
+using System.Threading;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal
+{
+    public class CoalescingScheduler : PipeScheduler
+    {
+        private static readonly WaitCallback _doWorkCallback = s => ((CoalescingScheduler)s).DoWork();
+
+        private readonly ConcurrentQueue<Action> _actions = new ConcurrentQueue<Action>();
+        private readonly ConcurrentQueue<Work> _workItems = new ConcurrentQueue<Work>();
+
+        private readonly object _workSync = new object();
+        private bool _doingWork;
+
+        public virtual void Schedule(Action action)
+        {
+            _actions.Enqueue(action);
+            TriggerWork();
+        }
+
+        public override void Schedule<T>(Action<T> action, T state)
+        {
+            var work = new Work
+            {
+                CallbackAdapter = (c, s) => ((Action<T>)c)((T)s),
```

Comment:
If the adapter is the same for every work item, why take the space to store the delegate?

Reply:
We won't need this once state goes back to object.

Technically the CallbackAdapter changes as `T` does. My understanding is that the `Action<object, object>` assigned here is static but it's static per `T` meaning a new `Action<object, object>` gets initialized the first time this method gets called with a given `T`. It should be the equivalent to what we do [here](https://github.com/aspnet/KestrelHttpServer/blob/f52771d0e9681798d8fbd71866a113d1053fb31d/src/Kestrel.Transport.Libuv/Internal/LibuvThread.cs#L420).